driver-pulsar/src/main/java/io/nosqlbench/driver/pulsar/ops/PulsarConsumerOp.java

```diff
@@ -147,10 +148,16 @@ public void run(Runnable timeTracker) {
                 }
 
                 if (!payloadRttTrackingField.isEmpty()) {
-                    // TODO: Extract the extractedSendTime from the payload and convert it to a long.
-                    long extractedSendTime = 0L;
-                    long delta = System.currentTimeMillis()-extractedSendTime;
-                    payloadRttHistogram.update(delta);
+                    String avroDefStr = pulsarSchema.getSchemaInfo().getSchemaDefinition();
+                    org.apache.avro.Schema avroSchema =
+                            AvroUtil.GetSchema_ApacheAvro(avroDefStr);
+                    org.apache.avro.generic.GenericRecord avroGenericRecord =
+                            AvroUtil.GetGenericRecord_ApacheAvro(avroSchema, message.getData());
+                    if (avroGenericRecord.hasField(payloadRttTrackingField)) {
+                        long extractedSendTime = (Long)avroGenericRecord.get(payloadRttTrackingField);
+                        long delta = System.currentTimeMillis() - extractedSendTime;
+                        payloadRttHistogram.update(delta);
+                    }
                 }
```

Comment:
There is substantial initialization logic here in the case that schema has to be loaded. This should be pushed towards the front of the activity's lifecycle. I'll merge it now so you can use the code, but it should get an "init vs run fast" optimization, as schema realization is likely to be non-trivial.

Reply:
Also it would have been better to use a message header instead of putting this field into the payload.
AVRO adds overhead and here we are measuring latency

Reply:
my understanding is that this feature was already implemented using "e2e_starting_time"

Reply:
This is not the same as `e2e_starting_time` AFAIK because that would require CNDB to extract the embedded start timestamp payload from the CQL INSERT and add it as a property to the pulsar message. `payloadRtt` doesn't require that kind of munging in theory (in practice, I never implemented a generic "map CQL types to AVRO types" method and instead pull out a specific column name but `payloadRtt` is more generic on paper).